src/compiler/crystal/config.cr

```diff
@@ -34,8 +34,40 @@ module Crystal
       Time.unix(time).to_s("%Y-%m-%d")
     end
 
-    def self.default_target_triple
-      {{env("CRYSTAL_CONFIG_TARGET")}} || LLVM.default_target_triple
+    @@default_target : Crystal::Codegen::Target?
+
+    def self.default_target : Crystal::Codegen::Target
+      @@default_target ||= begin
+        target = Crystal::Codegen::Target.new({{env("CRYSTAL_CONFIG_TARGET")}} || LLVM.default_target_triple)
+
+        if target.linux?
+          # The statically linked linux binary runs as well on linux-gnu as
+          # on linux-musl, but the default target needs to match the C
+          # library used on the current system.
+          # This can be automatically detected from the output of `ldd --version`
+          # in order to use the appropriate environment target.
+          default_libc = target.gnu? ? "-gnu" : "-musl"
+
+          target = Crystal::Codegen::Target.new(target.to_s.sub(default_libc, "-#{runtime_libc}"))
```

Comment:
Is there a way to do the detection of the environment before creating the target?  It is probably not that much of an overhead but it seems like since you are using string detection you only need to make the target once.

```crystal
if ({{env("CRYSTAL_CONFIG_TARGET")}} || LLVM.default_target_triple).includes?("linux")
  Crystal::Codegen::Target.new(target.to_s.sub({"-gnu" => "-#{runtime_libc}", "-musl" => "-#{runtime_libc}"}))
else
  Crystal::Codegen::Target.new({{env("CRYSTAL_CONFIG_TARGET")}} || LLVM.default_target_triple)
end
```

Reply:
It's true, but by using `#linux?` we are also checking that the name occurs only in a specific part, and the check is case insensitive already due to `LLVM.normalize_triple(target_triple.downcase)`.

It's good enough as is IMO.

Reply:
I agree. This code is executed exactly once per compiler run. There is no point in optimizing out a little bit of string operations while repeating already implemented logic.

Reply:
That makes sense :)